### ISSUE TITLE
OpenSSL 3.1 beta1 release checklist

### ISSUE THREAD

**Issue by t8m** [openssl/openssl#19864]
**Checklist**
- [x] The code is functionally complete in regards to the particular release objectives as set by OMC and OTC.
- [x] There is no remaining refactoring required by OMC or OTC for the release.
- [x] There are no remaining API additions required for the release.
- [x] All issues [triaged](https://www.openssl.org/policies/technical/release-requirements.html#triage-process) as regressions on the development branch from which the release is to be done must have a milestone assigned.
- [x] All issues or pull requests with the milestone for the beta release are closed.
- [x] There are no outstanding untriaged Coverity issues.
- [x] Coveralls coverage has not decreased overall from the previous release.
- [x] The CI must pass on the tip of the development branch before the release commits are added to the tree, including the daily CI builds.
- [x] The tree must be frozen for at least 3 business days. No changes apart from regression or security fixes should be merged during the freeze.
- [x] For 2 days before the release there should be no changes to ensure the daily CI builds run on the development tree tip.
- [x] In case of the first beta release the OTC should explicitly approve that the source is ready for a release with a vote.

**Comment by t8m** [comment]
Problems:
 - We do not currently fully triage issues as regressions
 - Coveralls does not run against any other branch than master
 - The tree would have to be frozen on Monday for release on Thursday

**Comment by slontis** [comment]
What was the point of the alpha release when the beta follows it so closely?

**Comment by t8m** [comment]
> What was the point of the alpha release when the beta follows it so closely?

Yeah, I think we should delay it. The two-weeks alpha cycle is too short if there is only one alpha.

**Comment by t8m** [comment]
**OTC is OK with skipping the requirement to run Coverity before the beta release, but we will attempt to get it run before the final release.**

**Comment by t8m** [comment]
**OTC would accept a local run of lcov to compare the coverage with 3.0 branch if using Coveralls would be impossible or too difficult.**

**Comment by paulidale** [comment]
Coverity will be clean on the 3.1 branch once #19900 is merged.

**Comment by romen** [comment]
@t8m when we discussed in the meeting the item

> - [x]  All issues or pull requests with the milestone for the beta release are closed.

we somehow "cheated" as there are currently no issues assigned with the 3.1 milestone.
At the same time, we do know that although not tracked as github issue, there is the issue of the broken pyca tests for the 3.1 branch.

I am not sure if we reached consensus on how to deal with it, while we wait for pyca to merge fixes:
- should we ignore the issue, even though we have failing CI?
- should we ignore the issue, and disable pyca-tests from the CI as we await external resolution?
- should we invest resources in trying to speed up ht development and merge of fixes in pyca, so that we can pass pyca-tests before beta1?

Did I miss if OTC reached consensus on how to deal with failing CI and pyca-tests before the 3.1 beta release?

---

Still related to 3.1 and pyca, in previous discussions @slontis did warn that we are likely to lose coverage on our codebase from the pyca-tests, if we start tracking the `main` branch of `pyca/cryptography` rather than the `38.0.x` branch, as `main` switched to rust-based implementations for some functionality previously backed by OpenSSL.
As backports to the `38.0.x` branch are not necessarily guaranteed upstream (even though so far the `pyca/cryptography` maintainers did kindly backport proposed fixes), what are our plans to deal with it?

---

So, in conclusion, do we need an explicit OTC decision on either of these 2 related `pyca-tests` issues before the `3.1.0-beta1` release?

**Comment by paulidale** [comment]
Any drop in coverage will require a decision.  Let's see if it happens first.
I'm less sure about the pyca external test failing.

The 3.1 milestones are: FIPS 140-3 and back porting performance enhancements.  I consider both to be "done" currently.  The associated PRs should have had the 3.1 milestone but didn't.  So a process oversight rather than "cheating".

**Comment by romen** [comment]
> So a process oversight rather than "cheating".

Sorry, yes, I was trying to be light-hearted in my use of "cheating", but I completely agree with your assessment, thanks for clarifying that.

---

I think we still might have hasted too much on checking the box next to

> - [x] All issues or pull requests with the milestone for the beta release are closed.

As we should probably have an issue with the 3.1 milestone to track "failing external pyca-tests for the 3.1 branch", as that is a blocker for

> - [ ] The CI must pass on the tip of the development branch before the release commits are added to the tree, including the daily CI builds.

I believe we either need to prioritize a resolution for the "failing external pyca-tests for the 3.1 branch" issue, or ask the OTC to take a decision about explicitly making an exception for pyca-tests for this beta1 release.

---

> Any drop in coverage will require a decision. Let's see if it happens first.
> I'm less sure about the pyca external test failing.

If we are forced to switch to a commit in `main@pyca-cryptography` and abandon the `38.0.x` branch, does Coverall detect any change in coverage that would require us to be proactive about it to meet the beta1 release date projected in the OTC pre-approval vote?
I have no idea if Coverall is only running our own tests to evaluate coverage, or if external tests are also part of the Coverall builds, and what is the impact of the changes in `main@pyca-cryptography` on that metric.

**Comment by romen** [comment]
This also has implications if @t8m evaluates it is not feasible to use Coverall to measure 3.1 and 3.0, resorting for local runs of `lcov` to satisfy the checklist requirement. Should the `lcov` measurements used in this assessment include external tests or not?

**Comment by t8m** [comment]
There is no pyca test failing on 3.1 branch. There is master branch only failure.

**Comment by t8m** [comment]
And BTW - we currently do not run pyca tests in the coveralls build so any change in pyca submodule does not affect the coverage in coveralls.

**Comment by t8m** [comment]
Instead of trying to make coveralls work, I've manually generated coverage results against builds of openssl-3.0.0, current openssl-3.0, and current openssl-3.1 checkouts. Results are below:

| checkout | openssl-3.1 | openssl-3.0 | openssl-3.0.0 |
| -------- | --------- | --------- | --------- |
| lines | 66.9 % | 66.8 % | 66.2 % |
| functions | 78.5 % | 78.3 % | 77.4 % |

**Comment by t8m** [comment]
The checklist is complete. I am going ahead with the release.
